### ISSUE TITLE
Tailscaled tells systemd that it is ready before its ip address is bindable

### ISSUE THREAD

**Issue by pvsr** [tailscale/tailscale#11504]
### What is the issue?

Originally reported in https://github.com/tailscale/tailscale/issues/3340#issuecomment-1790597859. That issue is about delaying startup of services that bind to a tailscale address using systemd. The suggested solution is to add:
```
[Unit] 
After=tailscaled.service
```
to ensure that the address is available before dependent services start, but this isn't working as it should because the daemon is reporting that it's ready, and therefore that services with `After=tailscaled.service` can start, before the IP address is actually available to bind. So usually when I reboot my NixOS server, all of the services I have configured with `After=tailscaled.service` will fail with errors about the tailscale IP address they're trying to bind not being available. By the time I find the command to restart all of the affected services in my shell history tailscaled has finished starting up and it works fine. But it's annoying to have to remember to do this after every reboot, or to add a hacky `ExecStartPre=/bin/sleep 10` workaround to each affected service. And if you were to bind sshd to a tailscale IP and it failed to start due to this bug, you could be left with an inaccessible system.

edit: another workaround, described [here](https://github.com/tailscale/tailscale/issues/3340#issuecomment-1021817151), is to allow binding to non-local ip addresses. I'm sure that works fine, but it's a relatively obscure global setting that is documented as breaking some applications so definitely not an ideal solution.

### Steps to reproduce

This happens consistently on my NixOS server and can probably be reproduced on any systemd installation on bare metal where the network takes some time to come online. Configure any service (e.g. sshd, nginx) to listen to a tailscale IP, add `After=tailscaled.service` to the systemd unit as suggested in the other issue, and reboot.

### Are there any recent changes that introduced the issue?

I believe this is a regression introduced in https://github.com/tailscale/tailscale/commit/e8cc78b1afdd4b7358c39b8cfbfcd7ae7faf3bd3 or earlier. [This line](https://github.com/tailscale/tailscale/blob/b104688e04797162fb854eb42158f983d05dc6cc/ipn/ipnserver/server.go#L541) tells systemd that the unit is ready, but the code right before it was refactored to start LocalBackend asynchronously and left `systemd.Ready()` running immediately after the async work kicks off. It should probably be moved to somewhere inside that async backend startup code.

### OS

Linux

### OS version

NixOS 23.11

### Tailscale version

1.58.2

**Comment by nagisa** [comment]
FWIW it is probably a good idea to set up such services so that they attempt restarting themselves anyway. Especially things like sshd, a crash of which may lead to inaccessible system.

**Comment by cdhowie** [comment]
We are hitting this issue. To work around it, we've applied the following changes to all of our systems via Puppet:

Create a file `/usr/local/sbin/tailscale-wait-for-ip` with the following content and mark it executable:

```
#!/bin/sh

echo Waiting for tailscale0 to get an IP address...
for i in 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15; do
        if ip addr show dev tailscale0 | grep -q 'inet '; then break; fi
        echo $i
        sleep 1
done
```

Put the following in `/etc/systemd/system/tailscaled.service.d/override.conf`:

```
[Service]
ExecStartPost=/usr/local/sbin/tailscale-wait-for-ip
```

This assumes the tailscale interface name is `tailscale0`. It's a bit of a kludge but it seems to work well enough.

**Comment by QuinnyPig** [comment]
This is helpful! I have nginx proxying traffic from a public-facing server over the tailnet, and nginx bombs out on reboot because the tailnet hasn't settled yet. Glad I found this...

**Comment by simonwiles** [comment]
imo, the cleanest way is to add this in `/etc/systemd/system/tailscaled.service.d/override.conf`:
```
ExecStartPost=timeout 60s bash -c 'until tailscale status --peers=false; do sleep 1; done'
```

then other services can just:
```
[Unit]
...
After=tailscaled.service
```

as they probably should anyway.

**Comment by lgrn** [comment]
> imo, the cleanest way is to add this in `/etc/systemd/system/tailscaled.service.d/override.conf`

Also good to know that there's a built in way to do this, you don't have to create the file manually:

```
systemctl edit tailscaled
```

This will open your editor, then as instructed by the comments, put this between the two comment blocks:

```
[Service]
ExecStartPost=timeout 60s bash -c 'until tailscale status --peers=false; do sleep 1; done'
```

Save the file, then run:

```
systemctl cat tailscaled
```

...to confirm that the override is there at the bottom.

Then just do `systemctl daemon-reload` to apply the change like usual.

**Comment by 0xFFA500** [comment]
> imo, the cleanest way is to add this in `/etc/systemd/system/tailscaled.service.d/override.conf`:
> 
> ```
> ExecStartPost=timeout 60s bash -c 'until tailscale status --peers=false; do sleep 1; done'
> ```
> 
> then other services can just:
> 
> ```
> [Unit]
> ...
> After=tailscaled.service
> ```
> 
> as they probably should anyway.



> > imo, the cleanest way is to add this in `/etc/systemd/system/tailscaled.service.d/override.conf`
> 
> Also good to know that there's a built in way to do this, you don't have to create the file manually:
> 
> ```
> systemctl edit tailscaled
> ```
> 
> This will open your editor, then as instructed by the comments, put this between the two comment blocks:
> 
> ```
> [Service]
> ExecStartPost=timeout 60s bash -c 'until tailscale status --peers=false; do sleep 1; done'
> ```
> 
> Save the file, then run:
> 
> ```
> systemctl cat tailscaled
> ```
> 
> ...to confirm that the override is there at the bottom.
> 
> Then just do `systemctl daemon-reload` to apply the change like usual.

Brilliant, thanks guys!

I was pulling out my hair because I was having issues with my Traefik Docker container not starting correctly on boot because I'm binding it to the Tailscale address so that it keeps the ports available for my LAN

I used your examples to edit he Tailscale and Docker services to wait for the Tailscale to actually have addresses available

```ini
[Unit]
After=tailscaled.service
Wants=tailscaled.service
```

**Comment by strongoose** [comment]
Just cross-posting a pretty decent workaround to this I found [on reddit](https://www.reddit.com/r/Tailscale/comments/ubk9mo/comment/jia3pwn/) as I thought it might be useful to people who end up on this issue.

tldr;
```
After=sys-subsystem-net-devices-tailscale0.device tailscaled.service
```

**Comment by ldpr** [comment]
If possible could we get this bug re-evaluated?

I'm new to tailscale and had a lot of fun setting it up today and it's definitely making my life easier - until I restarted some VMs to test and everything broke. Having this fixed would be a massive QOL improvement, as it is now I'll have to undo tailscale and wait.
Thanks!

**Comment by LennyPenny** [comment]
> Just cross-posting a pretty decent workaround to this I found [on reddit](https://www.reddit.com/r/Tailscale/comments/ubk9mo/comment/jia3pwn/) as I thought it might be useful to people who end up on this issue.
> 
> tldr;
> 
> ```
> After=sys-subsystem-net-devices-tailscale0.device tailscaled.service
> ```

I just tried this and cannot report success: 
```
systemctl status sys-subsystem-net-devices-tailscale0.device                                                                                                                                    
● sys-subsystem-net-devices-tailscale0.device - /sys/subsystem/net/devices/tailscale0
     Loaded: loaded
     Active: active (plugged) since Mon 2026-01-05 18:57:54 CET; 5min ago
```
```
journalctl -fu myservice.service
Jan 05 18:58:02 xx myservice[1863]: level="info" msg="Starting single HTTP frontend" address="x.x.x.x:1234"
Jan 05 18:58:02 xx myservice[1863]: level="fatal" msg="listen tcp x.x.x.x:1234: bind: cannot assign requested address"
```
```
cat /etc/systemd/system/myservice.service                                                                                                                                                        
[Unit]
After=network-online.target local-fs.target tailscaled.service sys-subsystem-net-devices-tailscale0.device
```

Might need something to check if the address is actually bindable (this looks like the bind attempt is 8 seconds after the device activated)

**Comment by lf94** [comment]
Painful.

**Comment by bradfitz** [comment]
From the first commit that added sdnotify support in 2020 (7ea809897df1764ea420be2ff6ae58459b0e6902), the Ready signal has always been about when tailscaled itself is up and ready for LocalAPI commands, not when the "IP address is bindable". That was never its goal.

Maybe this needs documentation.

And/or maybe we need a second mechanism for people to wait for bindable. Maybe we could notify readiness on a different systemd target or something.

**Comment by bradfitz** [comment]
I sent https://github.com/tailscale/tailscale/pull/18574 as a potential solution for this.

**Comment by bradfitz** [comment]
And @raggi sent #18577 building on that to add a new systemd target type that does what some people assumed the existing one did.

**Comment by bradfitz** [comment]
Last two PRs are merged and will be in the next (non-patch) release.

**Comment by smlx** [comment]
@bradfitz could you please confirm which minor version ships this new target? I don't seem to be seeing it in v1.96.2 on Debian.

**Comment by reinob** [comment]
version 1.96.2 (installed from pkgs.tailscale.com/stable/debian) does have "tailscale wait" and "tailscale ip --assert".
(and they work ;-)

**Comment by SteveHawk** [comment]
The commands do work, however the `tailscale-online.target` systemd unit added in https://github.com/tailscale/tailscale/pull/18577 and mentioned in the [doc](https://tailscale.com/docs/reference/tailscale-cli#wait) doesn't seem to be included in the release.

I checked both the debian release:

```bash
$ dpkg -L tailscale
/etc
/etc/default
/etc/default/tailscaled
/lib
/lib/systemd
/lib/systemd/system
/lib/systemd/system/tailscaled.service
/usr
/usr/bin
/usr/bin/tailscale
/usr/sbin
/usr/sbin/tailscaled
```

And the [Static binaries](https://pkgs.tailscale.com/stable/#static):

```bash
$ wget https://pkgs.tailscale.com/stable/tailscale_1.96.2_amd64.tgz
$ ls tailscale_1.96.2_amd64/*
tailscale_1.96.2_amd64/tailscale  tailscale_1.96.2_amd64/tailscaled

tailscale_1.96.2_amd64/systemd:
tailscaled.defaults  tailscaled.service
```

`tailscale-online.target` is not there. Maybe the release build process needs to be updated to include the new target file?

**Comment by bradfitz** [comment]
@SteveHawk, whoops. I sent https://github.com/tailscale/tailscale/pull/19078

**Comment by SteveHawk** [comment]
@bradfitz Thanks!

A follow-up question: should the `tailscale-wait-online.service` be automatically enabled during update/installation? If so, it probably should also be included in the various postinst/prerm/postrm scripts for each distros (deb/rpm/nix, etc.), basically everywhere where `tailscaled.service` is being installed.

**Comment by bradfitz** [comment]
I'll let @raggi answer that, but he's currently frolicking on some epic adventure on the edge of even Starlink availability so we might have to wait a minute.
